pallets/collator-selection/src/lib.rs

```diff
@@ -564,22 +636,36 @@ pub mod pallet {
 
 			let candidates = Self::candidates();
 			let candidates_len_before = candidates.len();
-			let active_candidates = Self::kick_stale_candidates(candidates);
-			let active_candidates_len = active_candidates.len();
-			let result = Self::assemble_collators(active_candidates);
-			let removed = candidates_len_before - active_candidates_len;
+			let removed_candidate_ids = Self::kick_stale_candidates(candidates.clone());
+			let active_candidate_ids = candidates
+				.iter()
+				.filter_map(|x| {
+					if removed_candidate_ids.contains(&x.who) {
+						None
+					} else {
+						Some(x.who.clone())
+					}
+				})
+				.collect();
+			let result = Self::assemble_collators(active_candidate_ids);
 
 			frame_system::Pallet::<T>::register_extra_weight_unchecked(
-				T::WeightInfo::new_session(candidates_len_before as u32, removed as u32),
+				T::WeightInfo::new_session(
+					candidates_len_before as u32,
+					removed_candidate_ids.len() as u32,
+				),
 				DispatchClass::Mandatory,
 			);
+			// Build performance map for the currently active validatorset in the just started session
+			Self::reset_collator_performance();
 			Some(result)
 		}
 		fn start_session(_: SessionIndex) {
 			// we don't care.
 		}
 		fn end_session(_: SessionIndex) {
 			// we don't care.
+			// RAD: Should we do kicking on end_session instead?
```

Comment:
Should we do kicking on end_session instead?

There's no real difference in function, it might make it clearer though that we're using performance data of the just ended session.

Reply:
Resolving as no-action as this is a strictly optional change